fix(install): support glob paths with compressed firmware files

ihuguet opened the pull request:
This pull request changes...

## Changes

Glob paths for firmware was not working with compressed firmware files, only with uncompressed ones. Fix that.

At the same time, improve a bit the debug logs of firmware installation:
- different message if the file doesn't exist and if the file cannot be installed
- report only one message of "possible firmware missing" if it's not found in any of the posible paths, instead of once per posible path

Example of logs before the changes:
```
Firmware iwlwifi-QuZ-a0-hr-b0-72.ucode
dracut-install: stat(/lib/firmware/updates/5.14.0-200.el9.x86_64/iwlwifi-QuZ-a0-hr-b0-72.ucode) != 0
dracut-install: Possible missing firmware iwlwifi-QuZ-a0-hr-b0-72.ucode for kernel module iwlwifi
dracut-install: stat(/lib/firmware/updates/iwlwifi-QuZ-a0-hr-b0-72.ucode) != 0
dracut-install: Possible missing firmware iwlwifi-QuZ-a0-hr-b0-72.ucode for kernel module iwlwifi
dracut-install: stat(/lib/firmware/5.14.0-200.el9.x86_64/iwlwifi-QuZ-a0-hr-b0-72.ucode) != 0
dracut-install: Possible missing firmware iwlwifi-QuZ-a0-hr-b0-72.ucode for kernel module iwlwifi
dracut-install: stat(/lib/firmware/iwlwifi-QuZ-a0-hr-b0-72.ucode) != 0
dracut-install: Possible missing firmware iwlwifi-QuZ-a0-hr-b0-72.ucode for kernel module iwlwifi
dracut-install: Firmware iwlwifi-1000-5.ucode
dracut-install: stat(/lib/firmware/updates/5.14.0-200.el9.x86_64/iwlwifi-1000-5.ucode) != 0
dracut-install: Possible missing firmware iwlwifi-1000-5.ucode for kernel module iwlwifi
dracut-install: stat(/lib/firmware/updates/iwlwifi-1000-5.ucode) != 0
dracut-install: Possible missing firmware iwlwifi-1000-5.ucode for kernel module iwlwifi
dracut-install: stat(/lib/firmware/5.14.0-200.el9.x86_64/iwlwifi-1000-5.ucode) != 0
dracut-install: Possible missing firmware iwlwifi-1000-5.ucode for kernel module iwlwifi
dracut-install: dracut_install('/lib/firmware/iwlwifi-1000-5.ucode.xz', '/lib/firmware/iwlwifi-1000-5.ucode.xz', 0, 0, 1)
dracut-install: dracut_install ret = 0
dracut-install: cp '/lib/firmware/iwlwifi-1000-5.ucode.xz' '/var/tmp/dracut.awubz5/initramfs/lib/firmware/iwlwifi-1000-5.ucode.xz'
dracut-install: dracut_install ret = 0
dracut-install: dracut_install '/lib/firmware/iwlwifi-1000-5.ucode' OK
```

Example after the changes:
```
dracut-install: Firmware iwlwifi-QuZ-a0-jf-b0-72.ucode
dracut-install: firmware file '/lib/firmware/updates/5.14.0-200.el9.x86_64/iwlwifi-QuZ-a0-jf-b0-72.ucode' not found
dracut-install: firmware file '/lib/firmware/updates/5.14.0-200.el9.x86_64/iwlwifi-QuZ-a0-jf-b0-72.ucode.zst' not found
dracut-install: firmware file '/lib/firmware/updates/5.14.0-200.el9.x86_64/iwlwifi-QuZ-a0-jf-b0-72.ucode.xz' not found
dracut-install: firmware file '/lib/firmware/updates/iwlwifi-QuZ-a0-jf-b0-72.ucode' not found
dracut-install: firmware file '/lib/firmware/updates/iwlwifi-QuZ-a0-jf-b0-72.ucode.zst' not found
dracut-install: firmware file '/lib/firmware/updates/iwlwifi-QuZ-a0-jf-b0-72.ucode.xz' not found
dracut-install: firmware file '/lib/firmware/5.14.0-200.el9.x86_64/iwlwifi-QuZ-a0-jf-b0-72.ucode' not found
dracut-install: firmware file '/lib/firmware/5.14.0-200.el9.x86_64/iwlwifi-QuZ-a0-jf-b0-72.ucode.zst' not found
dracut-install: firmware file '/lib/firmware/5.14.0-200.el9.x86_64/iwlwifi-QuZ-a0-jf-b0-72.ucode.xz' not found
dracut-install: firmware file '/lib/firmware/iwlwifi-QuZ-a0-jf-b0-72.ucode' not found
dracut-install: firmware file '/lib/firmware/iwlwifi-QuZ-a0-jf-b0-72.ucode.zst' not found
dracut-install: firmware file '/lib/firmware/iwlwifi-QuZ-a0-jf-b0-72.ucode.xz' not found
dracut-install: Possible missing firmware iwlwifi-QuZ-a0-jf-b0-72.ucode for kernel module iwlwifi
dracut-install: Firmware iwlwifi-1000-5.ucode
dracut-install: firmware file '/lib/firmware/updates/5.14.0-200.el9.x86_64/iwlwifi-1000-5.ucode' not found
dracut-install: firmware file '/lib/firmware/updates/5.14.0-200.el9.x86_64/iwlwifi-1000-5.ucode.zst' not found
dracut-install: firmware file '/lib/firmware/updates/5.14.0-200.el9.x86_64/iwlwifi-1000-5.ucode.xz' not found
dracut-install: firmware file '/lib/firmware/updates/iwlwifi-1000-5.ucode' not found
dracut-install: firmware file '/lib/firmware/updates/iwlwifi-1000-5.ucode.zst' not found
dracut-install: firmware file '/lib/firmware/updates/iwlwifi-1000-5.ucode.xz' not found
dracut-install: firmware file '/lib/firmware/5.14.0-200.el9.x86_64/iwlwifi-1000-5.ucode' not found
dracut-install: firmware file '/lib/firmware/5.14.0-200.el9.x86_64/iwlwifi-1000-5.ucode.zst' not found
dracut-install: firmware file '/lib/firmware/5.14.0-200.el9.x86_64/iwlwifi-1000-5.ucode.xz' not found
dracut-install: firmware file '/lib/firmware/iwlwifi-1000-5.ucode' not found
dracut-install: firmware file '/lib/firmware/iwlwifi-1000-5.ucode.zst' not found
dracut-install: dracut_install('/lib/firmware/iwlwifi-1000-5.ucode.xz', '/lib/firmware/iwlwifi-1000-5.ucode.xz', 0, 0, 1)
dracut-install: dracut_install ret = 0
dracut-install: cp '/lib/firmware/iwlwifi-1000-5.ucode.xz' '/var/tmp/dracut.I6YIi0/initramfs/lib/firmware/iwlwifi-1000-5.ucode.xz'
dracut-install: dracut_install ret = 0
dracut-install: dracut_install '/lib/firmware/iwlwifi-1000-5.ucode.xz' OK
```
## Checklist
- [x] I have tested it locally
- [ ] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

naggety commented:
I have tested the changes and I think they work fine.

Tried to run all tests, but MULTINIC-NM fails. However, it also fails without my changes.

About updating documentation and adding test code: I didn't see anything related to this that I can update. If there is something, please indicate where and what I should do.

LaszloGombos commented:
@naggety 

> Tried to run all tests, but MULTINIC-NM fails. However, it also fails without my changes.

Indeed, this is a known existing issue - https://github.com/dracutdevs/dracut/issues/2087

nabijaczleweli commented on src/install/dracut-install.c:
```diff
@@ -1367,81 +1367,95 @@ static int install_all(int argc, char **argv)
-                                                 kmod_module_get_name(mod));
-                                }
+                STRV_FOREACH(fwdir, firmwaredirs) {
+                        if (install_firmware_fullpath(*fwdir, fwname, "") == 0) {
+                                found = true;
+                                if (!is_glob)
+                                        continue;
+                        }
+                        if (install_firmware_fullpath(*fwdir, fwname, ".zst") == 0) {
+                                found = true;
+                                if (!is_glob)
+                                        continue;
+                        }
+                        if (install_firmware_fullpath(*fwdir, fwname, ".xz") == 0) {
+                                found = true;
+                                if (!is_glob)
+                                        continue;
```
The approach looks sound; I think it may make more sense to roll this loop up tho:
```c
static const char * const * firmware_suffixes[] = {"", ".zst", ".xz", NULL};
// ...
                STRV_FOREACH(fwdir, firmwaredirs) {
                        STRV_FOREACH(suff, firmware_suffixes) {
                                if (install_firmware_fullpath(*fwdir, fwname, suff) == 0) {
                                        found = true;
                                        if (!is_glob)
                                                continue;
                                }
                        }
                }
```

naggety replied:
Can do that. But maybe is a good moment to ask: do we need to install same firmware from different folders? i.e. from /lib/firmware/updates and also from /lib/firmware. Currently that's what is done, and I maintained it, but is it useful?

nabijaczleweli replied:
well, we need to correctly mirror the kernel's loading pattern – drivers/base/firmware_loader/main.c – the kernel searches for firmware A in all paths, then for firmware A.zst in all paths, then for firmware A.xz in all paths

if you invert the loop stacking to be `for(suff : suffixes) for(path : paths) if(install(path, firmware, suff) == ok) break 2;` then we could theoretically early-exit

nabijaczleweli commented on src/install/dracut-install.c:
```diff
@@ -1367,81 +1367,95 @@ static int install_all(int argc, char **argv)
+
+                if (access(fw, F_OK) != 0) {
+                        log_debug("stat(%s) != 0", fw);
+                        fail++;
+                        continue;
+                }
+
+                if (dracut_install(fw, fw, false, false, true) == 0) {
+                        log_debug("dracut_install '%s' OK", fw);
+                        ok++;
+                } else {
+                        log_debug("dracut_install '%s' FAIL", fw);
+                        fail++;
```
i mean, realistically, this isn't an issue, but this *is* a very archaic style, and you *could*, nominally, overflow ok and fail all the way back to zero on platforms where sizeof(size_t) > sizeof(int); why not make them both bools?

naggety replied:
We don't need to know the count for anything, so yes, it makes more sense to use bools. I will change it.

johannbg commented:
Reviewers remember to resolve conversation when they are done. 
@naggety minor nitpick can you change your commit message to fix(install): support glob paths with compressed firmware files as well as drop that sign-off reference since we dont do sign-offs in the project, it's a linux kernel thing and arguably practically useless one at that.

ihuguet committed:
fix(install): support glob paths with compressed firmware files

Glob paths for firmware was not working with compressed
firmware files, only with uncompressed ones. Fix that.

At the same time, improve a bit the debug logs of firmware
installation:
- different message if the file doesn't exist and if the file
  cannot be installed
- report only one message of "possible firmware missing" if it's
  not found in any of the posible paths, instead of once per
  posible path

nabijaczleweli commented on src/install/dracut-install.c:
```diff
@@ -1367,82 +1367,89 @@ static int install_all(int argc, char **argv)
-
-                        _asprintf(&fwpath, "%s/%s", *q, value);
+                fwname = kmod_module_info_get_value(l);
+                is_glob = strstr(fwname, "*") || strstr(fwname, "?") || strstr(fwname, "[");
```
`strpbrk(fwname, "*?[")`

naggety replied:
Done

naggety commented:
Hi, is there anything else needed here?

johannbg commented:
@naggety just minor linter issue btw did you file upstream kernel pr to check the path validation preference is well valid :smile:

naggety commented:
@johannbg fixed the lint issue. I haven't sent any patch yet to upstream, I didn't find the time to work on it yet.

nabijaczleweli commented:
cringe bot (dutch)

naggety commented:
Hi again. Is there something I have to do here in order to move it forward?

johannbg commented:
> Hi again. Is there something I have to do here in order to move it forward?

@naggety yes a) resolve conversation that have been resolved and b) resolve the merge conflict.
